Samples/src/VisualDebugger/websrc/app.ts

```diff
@@ -0,0 +1,304 @@
+import "./css/main.css";
+import * as signalR from "@aspnet/signalr";
+import * as chart from "chart.js";
+
+//#region Serialization contract
+
+type State = {
+    real: number,
+    imaginary: number,
+    magnitude: number,
+    phase: number
+}[];
+
+//#endregion
+
+//#region HTML elements
+
+const divOperations: HTMLDivElement = document.querySelector("#divOperations");
+const olOperations: HTMLOListElement = document.querySelector("#olOperations");
+const btnStepIn: HTMLButtonElement = document.querySelector("#btnStepIn");
+const btnStepOver: HTMLButtonElement = document.querySelector("#btnStepOver");
+const btnPrevious: HTMLButtonElement = document.querySelector("#btnPrevious");
+const canvas: HTMLCanvasElement = document.querySelector("#chartCanvas");
+const chartContext = canvas.getContext("2d");
+
+//#endregion
+
+const operations: HTMLLIElement[] = [];
+
+type Snapshot = {
+    state: State,
+    lastOperation: HTMLLIElement,
+    nextOperation: HTMLLIElement
+};
+
+type History = {
+    snapshots: Snapshot[],
+    position: number
+};
+
+const history: History = {
+    snapshots: [],
+    position: -1
+};
+
+const stateChart = new chart.Chart(chartContext, {
+    type: "bar",
+    data: {
+        labels: [],
+        datasets: [
+            {
+                data: [],
+                label: "Real",
+				backgroundColor: "#ff0000",
+				borderColor: "#ff0000",
+            },
+            {
+                data: [],
+                label: "Imag",
+				backgroundColor: "#0000ff",
+				borderColor: "#0000ff",
+            }
+        ]
+    },
+    options: {
+        responsive: true,
+        maintainAspectRatio: false,
+        scales: {
+            yAxes: [
+                {
+                    ticks: {
+                        suggestedMin: -1,
+                        suggestedMax: 1
+                    }
+                }
+            ]
+        }
+    }
+});
+
+function updateChart(state: State) {
```

Comment:
It would be nice if the x-axis showed the base vector as the label, is that possible?

Reply:
Yes, they should be there - are they not working? This is what I see:

![Basis Vectors](https://user-images.githubusercontent.com/33814365/63982986-9cadcc00-ca79-11e9-8dd2-7e1d084b0d81.png)

Reply:
Ah! You're right. I was expecting them to be in the x-axis line (i.e. on the `0`); further more, they actually do show but they are mostly covered by the scrollbar:

![image](https://user-images.githubusercontent.com/24311286/64479603-54c03080-d16e-11e9-923c-ca7f52d9b677.png)

It's ok if we leave them at the bottom, but could we push them a little higher so they are not blocked by the scrollbar?

Further, allocating new qubits is not showing the new basis state after the allocation, i.e.:
![image](https://user-images.githubusercontent.com/24311286/64479660-d7e18680-d16e-11e9-8450-dfc42abb2ffb.png)

Is that something we could add?

Reply:
Hmm, I don't see the bottom scrollbar even when I make my window small. What browser are you using?

Reply:
It seems like the `OnAllocateQubits` handler is called just before the qubits are actually allocated - so the state that I can get in the handler is out of date. For example after the third allocation, the state would show only 4 basis vectors, but after the first H it would show 8 vectors.

@cgranade Is it possible for the simulator to invoke `OnAllocateQubits` after calling `manager.Allocate`?

Reply:
@marshallsa, OnOperationStart is called *before* the operation is called, `onAllocate` should remain consistent with that.
Are you saying that before calling 'H', there are only 4 basis state? that seems odd, there might be something else going on, by the time 'H' is highlighted (i.e. after `Allocate` is called) there should already be 8 basis state vectors.

Reply:
No, I think you're right that there are 8 basis vectors by the time `OnOperationStart` is called for H, but I was trying to show all 8 of them on the last "allocate 1 qubit" line instead of the first "H" line. Do you prefer showing 4 basis states on the last allocate line and then at the start of the first H line (before H actually runs), it shows 8?

Reply:
I think it's ok to be consistent, so when the last Allocation is highlighted then there should be 4, when H gets highlighted there should be 8. Again, the instruction listed is what is going to be executed, not what was already executed.

Reply:
That makes sense - allocation should work that way now. I also changed the green checkmark to a blue arrow like the rest of the operations so it's clear that the allocation is the next thing to execute. Sorry for the confusion!